testing/integration_testing/restore_secrets.py

```diff
@@ -0,0 +1,137 @@
+parser.add_argument("--passphrase", help="The passphrase itself.")
+parser.add_argument("--passphrase_file", help="Path to file with passphrase.")
+parser.add_argument("--repo_dir", default=os.getcwd(), help="Path to SDK Repo")
+args = parser.parse_args()
+
+
+def main():
+  repo_dir = args.repo_dir
+  # The passphrase is sensitive, do not log.
+  if args.passphrase:
+    passphrase = args.passphrase
+  elif args.passphrase_file:
+    with open(args.passphrase_file, "r") as f:
+      passphrase = f.read()
+  else:
+    raise ValueError("Must supply passphrase or passphrase_file arg.")
+
+  secrets_dir = os.path.join(repo_dir, "scripts", "gha-encrypted")
+  encrypted_files = _find_encrypted_files(secrets_dir)
+  print("Found these encrypted files:\n%s" % "\n".join(encrypted_files))
+
+  for path in encrypted_files:
+    if "google-services" in path or "GoogleService" in path:
+      print("Encrypted Google Service file found: %s" % path)
+      # We infer the destination from the file's directory, example:
+      # /scripts/gha-encrypted/auth/google-services.json.gpg turns into
+      # /<repo_dir>/auth/integration_test/google-services.json
+      api = os.path.basename(os.path.dirname(path))
+      file_name = os.path.basename(path).replace(".gpg", "")
+      dest_path = os.path.join(repo_dir, api, "integration_test", file_name)
+      decrypted_text = _decrypt(path, passphrase)
+      with open(dest_path, "w") as f:
+        f.write(decrypted_text)
+      print("Copied decrypted google service file to %s" % dest_path)
+
+  print("Attempting to patch Dynamic Links uri prefix.")
+  uri_path = os.path.join(secrets_dir, "dynamic_links", "uri_prefix.txt.gpg")
+  uri_prefix = _decrypt(uri_path, passphrase)
+  dlinks_project = os.path.join(repo_dir, "dynamic_links", "integration_test")
+  _patch_main_src(dlinks_project, "REPLACE_WITH_YOUR_URI_PREFIX", uri_prefix)
+
+  print("Attempting to patch Messaging server key.")
+  server_key_path = os.path.join(secrets_dir, "messaging", "server_key.txt.gpg")
+  server_key = _decrypt(server_key_path, passphrase)
+  messaging_project = os.path.join(repo_dir, "messaging", "integration_test")
+  _patch_main_src(messaging_project, "REPLACE_WITH_YOUR_SERVER_KEY", server_key)
+
+
+def _find_encrypted_files(directory_to_search):
+  """Returns a list of full paths to all files encrypted with gpg."""
+  encrypted_files = []
+  for prefix, _, files in os.walk(directory_to_search):
+    for relative_path in files:
+      if relative_path.endswith(".gpg"):
+        encrypted_files.append(os.path.join(prefix, relative_path))
+  return encrypted_files
+
+
+def _decrypt(encrypted_file, passphrase):
+  """Generates a decrypted file with same path minus the '.gpg' extension."""
+  print("Decrypting %s" % encrypted_file)
+  # Note: if setting check=True, be sure to catch the error and not rethrow it
+  # or print a traceback, as the message will include the passphrase.
+  result = subprocess.run(
+      args=[
+          "gpg",
```

Comment:
Should we install gpg in`install_prereqs` scripts? Wondering if it makes sense to just install the python gpg library instead of invoking system package manager to install this?

Reply:
We should install gpg somewhere, assuming it's not already available on GHA machines. Though it's worth noting that we only want to install it for ourselves, and on CI, not for external contributors or users, since it's used for decrypting secrets that only we should have access to.

Looking at the Python module, it appears to be a wrapper for the gpg tool itself?
"Apart from a recent-enough version of Python, in order to use this module you need to have access to a compatible version of the GnuPG executable."

In which case we would still need it installed.
https://pythonhosted.org/python-gnupg/

Reply:
Looked into it, and GHA documentation states that GPG is installed on MacOS and Linux, but it's not listed for Windows.

Though I tested it out, and found GPG installed and on the path for all three platforms.

Reply:
This workflow will run install_prereqs for all platforms, yes? (once ios/android are moved to python). We could have the prereqs scripts *verify* some prereqs and error out if they aren't present. e.g. gpg. So if the runners change we know exactly why by seeing the prereqs script output.